pkg/oci/oci.go

```diff
@@ -478,16 +490,141 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
+	}
+
+	imageDigest := desc.Digest.String()
+	signatureTag, err := craftSignatureTag(imageDigest)
+	if err != nil {
+		return "", "", fmt.Errorf("crafting signature tag: %w", err)
+	}
+
+	repo, err := newRepository(imageRef, authOpts)
+	if err != nil {
+		return "", "", fmt.Errorf("creating repository: %w", err)
+	}
+
+	_, signatureBytes, err := oras.FetchBytes(ctx, repo, signatureTag, oras.DefaultFetchBytesOptions)
+	if err != nil {
+		return "", "", fmt.Errorf("getting signature bytes: %w", err)
+	}
+
+	signatureManifest := &ocispec.Manifest{}
+	err = json.Unmarshal(signatureBytes, signatureManifest)
+	if err != nil {
+		return "", "", fmt.Errorf("decoding signature manifest: %w", err)
+	}
+
+	layers := signatureManifest.Layers
+	expectedLen := 1
+	layersLen := len(layers)
+	if layersLen != expectedLen {
+		return "", "", fmt.Errorf("wrong number of signature manifest layers: expected %d, got %d", expectedLen, layersLen)
+	}
+
+	layer := layers[0]
+	expectedMediaType := "application/vnd.dev.cosign.simplesigning.v1+json"
+	if layer.MediaType != expectedMediaType {
+		return "", "", fmt.Errorf("wrong layer media type: expected %s, got %s", expectedMediaType, layer.MediaType)
+	}
+
+	signature, ok := layer.Annotations["dev.cosignproject.cosign/signature"]
+	if !ok {
+		return "", "", fmt.Errorf("no signature in layer")
+	}
+
+	payloadTag := layer.Digest.String()
+	// The payload is stored as a blob, so we fetch bytes from the blob store and
+	// not the manifest one.
+	_, payloadBytes, err := oras.FetchBytes(ctx, repo.Blobs(), payloadTag, oras.DefaultFetchBytesOptions)
+	if err != nil {
+		return "", "", fmt.Errorf("getting payload bytes: %w", err)
+	}
+
+	img := &payload.SimpleContainerImage{}
+	err = json.Unmarshal(payloadBytes, img)
+	if err != nil {
+		return "", "", fmt.Errorf("decoding payload: %w", err)
+	}
+
+	if img.Critical.Image.DockerManifestDigest != imageDigest{
+		return "", "", fmt.Errorf("payload digest does not correspond to image: expected %s, got %s", imageDigest, img.Critical.Image.DockerManifestDigest)
+	}
+
+	return signature, string(payloadBytes), nil
+}
+
+func verifyImage(ctx context.Context, image string, authOpts *AuthOptions) error {
+	sign, payload, err := getSigningInformation(ctx, image, authOpts)
+	if err != nil {
+		return fmt.Errorf("getting payload and signature: %w", err)
+	}
+
+	decodedSignature, err := base64.StdEncoding.DecodeString(sign)
+	if err != nil {
+		return fmt.Errorf("decoding signature: %w", err)
+	}
+
+	block, _ := pem.Decode([]byte(authOpts.PublicKey))
+	if block == nil {
+		return fmt.Errorf("decoding public key to PEM blocks")
+	}
+
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("parsing public key: %w", err)
+	}
+
+	verifier, err := signature.LoadVerifier(pub, crypto.SHA256)
+	if err != nil {
+		return fmt.Errorf("loading verifier: %w", err)
+	}
+
+	err = verifier.VerifySignature(bytes.NewReader(decodedSignature), strings.NewReader(payload))
```

Comment:
I don't understand why we need to pass the payload. The only useful info I see there is the `docker-manifest-digest` and we already verified it. Is it verified again?

```json
{
  "critical": {
    "identity": {
      "docker-reference": "172.28.128.1:5000/snapshot_socket"
    },
    "image": {
      "docker-manifest-digest": "sha256:4e0a8becac464f4cd37cc871a654639fba6b36c81c6841cede72140231447120"
    },
    "type": "cosign container image signature"
  },
  "optional": null
}
```

Reply:
Aahh with that payload we ask the verifier to verify with its pub key and the signature we are passing, the manifest with digest `docker-manifest-digest`. Makes sense. Thanks!